### PR TITLE
mgr/cephadm: add max-per-host to schedule multiple daemons on a single host

### PR DESCRIPTION
**Pull request by liewegas**
- for now, does not consider ports.. that is next
- refactors and simplifies the scheduling code to use fewer helpers and be more explicit
- drop the "avoid even numbers of monitors" logic that is IMO unnecessary and complicated

The next step is to handle port assignments...

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/python-common/ceph/tests/test_service_spec.py [hunk 1]
```diff
@@ -62,6 +62,7 @@ def test_parse_host_placement_specs(test_input, expected, require_network):
         ('3 data[1-3]', "PlacementSpec(count=3, host_pattern='data[1-3]')"),
         ('3 data?', "PlacementSpec(count=3, host_pattern='data?')"),
         ('3 data*', "PlacementSpec(count=3, host_pattern='data*')"),
+        ("count:2 max-per-host:4", "PlacementSpec(count=2, max_per_host=4)"),
```
Comment (sebastian-philipp) [inline src/python-common/ceph/tests/test_service_spec.py:65]: Would be great to have a special case for "just allow co-located daemons. I don't care about details" maybe `PlacementSpec(max_per_host=-1)` or so as `max_per_host` is a bit special for RGW.

File: src/pybind/mgr/cephadm/serve.py [hunk 2]
```diff
@@ -608,18 +608,18 @@ def virtual_ip_allowed(host):
                 daemons.append(sd)
 
         # remove any?
-        def _ok_to_stop(remove_daemon_hosts: Set[orchestrator.DaemonDescription]) -> bool:
-            daemon_ids = [d.daemon_id for d in remove_daemon_hosts]
+        def _ok_to_stop(remove_daemons: Set[orchestrator.DaemonDescription]) -> bool:
+            daemon_ids = [d.daemon_id for d in remove_daemons]
             assert None not in daemon_ids
             # setting force flag retains previous behavior, should revisit later.
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/serve.py:614]: while you're here: 

```suggestion
            # setting force flag retains previous behavior
```

File: src/pybind/mgr/cephadm/schedule.py [hunk 3]
```diff
@@ -216,18 +216,20 @@ def remove_daemon_hosts(self, host_pool: List[HostPlacementSpec]) -> Set[DaemonD
         return _remove_daemon_hosts
 
     def get_candidates(self) -> List[HostPlacementSpec]:
+        per_host = self.spec.placement.get_num_per_host()
+
         if self.spec.placement.hosts:
-            return self.spec.placement.hosts
-        elif self.spec.placement.label:
+            return self.spec.placement.hosts * per_host
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/schedule.py:207]: That doesn't work. this has to be more complicated: 

```yaml
placement:
  hosts:    
  - mon-host-a=a
  - mon-host-b=b
  - mon-host-c=c
  max_per_host=3
```
would create garbage. 

For explicit host placement, we already have a way to allow explicit co-location:

```yaml
placement:
  hosts:    
  - mon-host-a=a
  - mon-host-a=b
  - mon-host-a=c
```

I'd be :+1: for making `max_per_host` incompatible with explicit host placement for now

File: src/python-common/ceph/tests/test_service_spec.py [hunk 4]
```diff
@@ -74,6 +75,8 @@ def test_parse_placement_specs(test_input, expected):
         ("host=a host*"),
         ("host=a label:wrong"),
         ("host? host*"),
+        ('host=a max-per-host:0'),
+        ('host=a max-per-host:-10'),
```
Comment (sebastian-philipp) [inline src/python-common/ceph/tests/test_service_spec.py:79]: Would be great to prevent this as well:

```yaml
placement:
  hosts:
  - hosta=a
  - hosta=b
  max-per-host=4
```

File: src/pybind/mgr/cephadm/tests/test_scheduling.py [hunk 5]
```diff
@@ -349,6 +349,17 @@ class NodeAssignmentTest(NamedTuple):
...
+        NodeAssignmentTest(
+            'mgr',
+            PlacementSpec(host_pattern='*', max_per_host=2),
+            'host1 host2 host3'.split(),
+            [
+                DaemonDescription('mgr', 'a', 'host1'),
+                DaemonDescription('mgr', 'b', 'host2'),
+            ],
+            ['host1', 'host2', 'host3', 'host1', 'host2', 'host3']
+        ),
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/tests/test_scheduling.py]: This trivial looking change is full of edge cases. 

```suggestion
        # all_hosts + max_per_host
        NodeAssignmentTest(
            'mgr',
            PlacementSpec(host_pattern='*', max_per_host=2),
            'host1 host2 host3'.split(),
            [
                DaemonDescription('mgr', 'a', 'host1'),
                DaemonDescription('mgr', 'b', 'host2'),
            ],
            ['host1', 'host2', 'host3', 'host1', 'host2', 'host3']
        ),
        # max_per_host + reduce count per host. This needs to remove MGRs d,e and f
        NodeAssignmentTest(
            'mgr',
            PlacementSpec(host_pattern='*', max_per_host=1),
            'host1 host2 host3'.split(),
            [
                DaemonDescription('mgr', 'a', 'host1'),
                DaemonDescription('mgr', 'b', 'host2'),
                DaemonDescription('mgr', 'c', 'host3'),
                DaemonDescription('mgr', 'd', 'host1'),
                DaemonDescription('mgr', 'e', 'host2'),
                DaemonDescription('mgr', 'f', 'host3'),
            ],
            ['host1', 'host2', 'host3']
        ),            
        # max_per_host + count
        NodeAssignmentTest(
            'mgr',
            PlacementSpec(host_pattern='*', count=4, max_per_host=2),
            'host1 host2 host3'.split(),
            [
                DaemonDescription('mgr', 'a', 'host1'),
                DaemonDescription('mgr', 'b', 'host2'),
                DaemonDescription('mgr', 'c', 'host3'),
            ],
            ['host1', 'host2', 'host3', 'host1']
        ),            
        # max_per_host + count + existsing daemons
        NodeAssignmentTest(
            'mgr',
            PlacementSpec(host_pattern='*', count=4, max_per_host=2),
            'host1 host2 host3'.split(),
            [
                DaemonDescription('mgr', 'a', 'host1'),
                DaemonDescription('mgr', 'b', 'host2'),
                DaemonDescription('mgr', 'c', 'host3'),
                DaemonDescription('mgr', 'd', 'host2'),
            ],
            ['host1', 'host2', 'host3', 'host2'],
        ),            
        # max_per_host + count + uneven existsing daemons
        NodeAssignmentTest(
            'mgr',
            PlacementSpec(host_pattern='*', count=4, max_per_host=2),
            'host1 host2 host3'.split(),
            [
                DaemonDescription('mgr', 'a', 'host1'),
                DaemonDescription('mgr', 'b', 'host1'),
                DaemonDescription('mgr', 'c', 'host3'),
                DaemonDescription('mgr', 'd', 'host3'),
            ],
            ['host1', 'host1', 'host3', 'host3'], # avoid relocation if possible
        ),
        # max_per_host + count + uneven existsing daemons on wrong hosts
        NodeAssignmentTest(
            'mgr',
            PlacementSpec(host_pattern='*', count=4, max_per_host=2),
            'host1 host2 host3'.split(),
            [
                DaemonDescription('mgr', 'a', 'host1'),
                DaemonDescription('mgr', 'b', 'host1'),
                DaemonDescription('mgr', 'c', 'host4'),
                DaemonDescription('mgr', 'd', 'host4'),
            ],
            ['host1', 'host1', 'host2', 'host3'],
        ),
        # max_per_host + count + spread out
        NodeAssignmentTest(
            'mgr',
            PlacementSpec(host_pattern='*', count=4, max_per_host=2),
            'host1 host2 host3'.split(),
            [
                DaemonDescription('mgr', 'a', 'host1'),
                DaemonDescription('mgr', 'b', 'host1'),
                DaemonDescription('mgr', 'c', 'host1'),
                DaemonDescription('mgr', 'd', 'host1'),
            ],
            ['host1', 'host1', 'host2', 'host3'],
        ),
```

File: src/pybind/mgr/orchestrator/_interface.py [hunk 6]
```diff
@@ -966,6 +966,14 @@ def service_name(self) -> str:
             return f'{daemon_type_to_service(self.daemon_type)}.{self.service_id()}'
         return daemon_type_to_service(self.daemon_type)
 
+    def get_host_placement(self) -> HostPlacementSpec:
+        return HostPlacementSpec(
+            hostname=self.hostname or '',
+            # FIXME: include the ip:port here?
+            network='',
+            name='',
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/_interface.py:974]: In case we already have HostPlacementSpecs in the ServiceSpec, we should get that info from there instead.

File: src/pybind/mgr/cephadm/schedule.py [hunk 7]
```diff
@@ -135,67 +135,52 @@ def place(self):
...
-        return self.scheduler.place(hosts, count)
+        # sort candidates into existing/used slots that already have a
+        # daemon, and others (the rest)
+        existing_active: List[HostPlacementSpec] = []
+        existing_standby: List[HostPlacementSpec] = []
+        others = candidates
+        for d in self.daemons:
+            hs = d.get_host_placement()
+            for i in others:
+                if str(i) == str(hs):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/schedule.py:146]:
```suggestion
                if i == hs:

File: src/pybind/mgr/cephadm/schedule.py [hunk 8]
```diff
@@ -135,67 +135,52 @@ def place(self):
...
+                    others.remove(i)
+                    if d.is_active:
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/schedule.py:148]: This looks like a followup PR to me. As it needs lots of unit tests as well

File: src/pybind/mgr/cephadm/schedule.py [hunk 9]
```diff
@@ -135,67 +135,52 @@ def place(self):
...
+        # Filter out hosts that don't have a public network assigned
+        # or don't allow virtual ips respectively
+        if self.filter_new_host:
+            old = others
+            others = [h for h in others if self.filter_new_host(h.hostname)]
+            for h in list(set(old) - set(others)):
+                if self.spec.service_type == 'ha-rgw':
+                    logger.info(
+                        f"Filtered out host {h.hostname} for ha-rgw. Could not verify host allowed virtual ips")
+            logger.debug('filtered %s down to %s' % (old, others))
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/schedule.py]: this is unfortunately completely untested. But we can't do anything about it right now.

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]
needs more tests

File: src/pybind/mgr/cephadm/schedule.py (hunk 7 above)
Comment (liewegas) [reply src/pybind/mgr/cephadm/schedule.py:146]: i tried that first but it didn't work.  do i need to create a __eq__ method or somethign?

Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/schedule.py:142]: works for me:

```python
In [2]: from  ceph.deployment.service_spec import HostPlacementSpec

In [3]: HostPlacementSpec('h', '', '') == HostPlacementSpec('h', '', '')
Out[3]: True

In [4]: HostPlacementSpec('h', '', '') == HostPlacementSpec('other', '', '')
Out[4]: False

In [5]: HostPlacementSpec('h', '', '') == HostPlacementSpec('h', '', 'othername')
Out[5]: False
```

Comment (liewegas) [reply src/pybind/mgr/cephadm/schedule.py]: hrm yeah me too... i must have been seeing something else when i was testing this early on
